### ISSUE TITLE
Merging configs in OmegaConf 2.0.0rc22

### ISSUE THREAD

**Issue by jonasteuwen** [omry/omegaconf#200]
In my model setup and training, I made a configuration example like this (holding values which I always require) (made shorter):

```python
from dataclasses import dataclass

@dataclass
class DefaultConfig:
    N_EPOCHS: int = 50
```

I would then want to extend the config for different use cases for instance, config.yaml is a yaml holding a 'DEBUG: true' key. 

>     default_cfg = OmegaConf.structured(DefaultConfig)
>     cfg = OmegaConf.merge(default_cfg, OmegaConf.load('config.yaml'))

If I do this I would expect that cfg now has merged the default and the config.yaml, but I get this error 

> omegaconf.errors.ConfigKeyError: Error setting DEBUG=True : Key 'DEBUG' not in 'DefaultConfig'
>         full_key: DEBUG
>         reference_type=Optional[DefaultConfig]
>         object_type=DefaultConfig

**Comment by omry** [comment]
Okay ,this is very intentional.
you can open DefaultConfig to arbitrary keys by extending Dict[str, Any], but in the general sense you probably want to embrace this type safety and structure your config differently.

Specifically for debug, you can put one of the following in your DefaultConfig.
```python
# default to False
DEBUG : bool  = False

# Default to unspecified
DEBUG : Optional[bool] = None

# No default, user must provide the value before runtime (Through merging or through command line overrides).
DEBUG : bool  = MISSING
```

**Comment by jonasteuwen** [comment]
Hello @omry, thanks your reply. One of my intentions is to prevent my DefaultConfig holding too many variables which will not always be used (for instance when using different neural networks as backends with different parameters). What would be the way enabling the type safety?

**Comment by omry** [comment]
I recommend that you add those things anyway.
The advantages are:

you can get static type checking for them.
you can get runtime type safety for them.
With a hierarchy of configs, you can make things cleaner.
for example, you can put aux things in a node under the main config.

Regardless of Structured configs, the real answer here is to use a hierarchy.

But if you want to keep to your wicked ways, try:
```python3
from typing import Dict, Any

@dataclass
class DefaultConfig(Dict[str, Any]):
    N_EPOCHS: int = 50
```

**Comment by jonasteuwen** [comment]
Thanks for your suggestion. If I would follow up on your first idea, and create configuration objects allowing both static typing and runtime type safety I still would have a lot of unused configuration variables loaded (when you do `cfg.pretty()`).

For instance, you might have a detection framework involving retina net, yolo, etc. These have different parameters and preferably you would have a cfg flag saying `NETWORK` and have these properties hierarchically loaded like `cfg.YOLO.any_var`, while you do not necessarily want to have all variables for the retina net loaded in the main configuration. But maybe this is just aesthetics but the configuration can become very big and hard to parse in your log files.

**Comment by omry** [comment]
try this:
```python
@dataclass
class ModelConf:
  ...

@dataclass
class ResnetModelConf(ModelConf):
  ...

@dataclass
class Config:
  model : ModelConf = MISSING
```

later in the code:
```python
cfg = OmegaConf.structured(Config)
# this
cfg = OmegaConf.merge(cfg, {"model" : ResnetConfig})
# or this:
cfg.model = ResnetConfig()
```
Now the runtime type of model is ResenetConfig. any manipulation (direct assignment, merge) would be validated against ResnetConfig,

**Comment by omry** [comment]
Closing, feel free to ask additional related questions on this issue or to open a new one for something else.
